modules/partners/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Number of blocks a referral lasts
+		#[pallet::constant]
+		type ReferralExpire: Get<Self::BlockNumber>;
+
+		/// Treasury's AccountId
+		#[pallet::constant]
+		type TreasuryAccount: Get<Self::AccountId>;
+
+		/// Max size of metadata string
+		#[pallet::constant]
+		type MaxMetadataLength: Get<u32>;
+
+		/// Provides current blocknumber
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Sub Account cant be generated, should not happen
+		SubAccountGenerationFailed,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Referral succesfully set to PartnerId
+		ReferralSet { who: T::AccountId, partner_id: PartnerId },
+		/// New Partner successfully registered
+		PartnerRegistered {
+			owner: T::AccountId,
+			partner_id: PartnerId,
+			partner_account: T::AccountId,
+		},
+		/// Partner's metadata is updated
+		PartnerMetadataUpdated { partner_id: PartnerId },
+	}
+
+	#[pallet::storage]
+	pub type NextId<T: Config> = StorageValue<_, PartnerId, ValueQuery>;
+
+	#[pallet::storage]
+	pub type PartnerMetadata<T: Config> =
+		StorageMap<_, Identity, PartnerId, BoundedVec<u8, T::MaxMetadataLength>, OptionQuery>;
+
+	#[pallet::storage]
+	pub type Referral<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, ReferralInfo<T::BlockNumber>, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn register_partner(
+			origin: OriginFor<T>,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// transfer registration fee to treasury
+			let treasury_account = T::TreasuryAccount::get();
+			<T as module::Config>::Currencies::transfer(
+				T::GetNativeCurrencyId::get(),
+				&who,
+				&treasury_account,
+				T::RegisterFee::get(),
+			)?;
+
+			Self::feeless_register_partner(who, metadata)
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn set_referral(origin: OriginFor<T>, partner: PartnerId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Referral::<T>::insert(
+				&who,
+				ReferralInfo {
+					partner_id: partner,
+					expiry: T::BlockNumberProvider::current_block_number().saturating_add(T::ReferralExpire::get()),
+				},
+			);
+
+			Self::deposit_event(Event::ReferralSet {
+				who,
+				partner_id: partner,
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn admin_register_partner(
+			origin: OriginFor<T>,
+			owner: T::AccountId,
+			metadata: BoundedVec<u8, T::MaxMetadataLength>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			Self::feeless_register_partner(owner, metadata)
+		}
+
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn update_partner(
```

Comment:
Maybe find a way to securely make this feeless, people probably would be more inclined to sign transaction if it was free